### PR TITLE
Hard-link local sources instead of symlinking them.

### PR DESCRIPTION
**Pull request by kyrofa**
This PR fixes LP: [#1590108](https://bugs.launchpad.net/snapcraft/+bug/1590108) by hard-linking local sources instead of symlinking them. If hard links aren't possible, it copies them instead.

### CODE REVIEW

File: snapcraft/internal/sources.py [hunk 1]
```diff
@@ -325,17 +326,26 @@ def provision(self, dst, clean_target=True, keep_zip=False):
...
+        if os.path.islink(self.source_dir) or os.path.isfile(self.source_dir):
             os.remove(self.source_dir)
-        elif (os.path.isdir(self.source_dir) and
-              not os.listdir(self.source_dir)):
-            os.rmdir(self.source_dir)
-        elif os.path.exists(self.source_dir):
-            raise EnvironmentError('Cannot pull to target {!r}'.format(
-                self.source_dir))
+        elif os.path.isdir(self.source_dir):
+            shutil.rmtree(self.source_dir)
```
Comment (kyrofa) [inline snapcraft/internal/sources.py]: Take particular note here: previously the code was very careful not to overwrite an existing `parts/<name>/src` directory or file. That logic is essentially impossible now that we're not symlinking, so I decided to assert ownership over `parts/<name>/` and just blow it away if it already exists. See any downside to that? Users shouldn't be putting things there...

File: snapcraft/_baseplugin.py [hunk 2]
```diff
@@ -144,6 +144,9 @@ def build(self):
         if os.path.exists(self.build_basedir):
             shutil.rmtree(self.build_basedir)
 
+        # FIXME: It's not necessary to ignore here anymore since it's now done
```
Comment (sergiusens) [inline snapcraft/_baseplugin.py]: please recreate an integration test with this scenario to confirm this

File: snapcraft/internal/sources.py (hunk 1 above)
Comment (sergiusens) [inline snapcraft/internal/sources.py]: ownership was always the case.
We do need to track source changes now though ;-)

Comment (kyrofa) [inline snapcraft/internal/sources.py]: > We do need to track source changes now though ;-)

I'm not sure what you mean. Are you saying something needs to happen here?

File: snapcraft/tests/test_sources.py [hunk 3]
```diff
@@ -384,51 +384,103 @@ def test_init_with_source_branch_and_tag_raises_exception(self):
...
+            os.stat(os.path.join('destination', 'dir', 'file')).st_nlink, 1)
+
+    def test_pull_ignores_snapcraft_specific_data(self):
+        # Make the snapcraft-specific directories
+        os.makedirs(os.path.join('src', 'parts'))
+        os.makedirs(os.path.join('src', 'stage'))
+        os.makedirs(os.path.join('src', 'prime'))
+
+        # Make the snapcraft.yaml and a built snap
+        open(os.path.join('src', 'snapcraft.yaml'), 'w').close()
```
Comment (sergiusens) [inline snapcraft/tests/test_sources.py]: ah, we might be missing .snapcraft.yaml here!

Comment (kyrofa) [inline snapcraft/tests/test_sources.py]: Ah, indeed we are! I'll go ahead and fix that here.

**Comment by sergiusens** [comment]
looks good, I am reaching EOD so will let @elopio pay attention to the detail in UTs ;-)

File: snapcraft/internal/sources.py (hunk 1 above)
Comment (sergiusens) [inline snapcraft/internal/sources.py]: El 01/07/16 a las 18:25, Kyle Fazzari escribió:
> In snapcraft/internal/sources.py
> <https://github.com/snapcore/snapcraft/pull/622#discussion_r69357549>:
> 
>>              os.remove(self.source_dir)
>> -        elif (os.path.isdir(self.source_dir) and
>> -              not os.listdir(self.source_dir)):
>> -            os.rmdir(self.source_dir)
>> -        elif os.path.exists(self.source_dir):
>> -            raise EnvironmentError('Cannot pull to target {!r}'.format(
>> -                self.source_dir))
>> +        elif os.path.isdir(self.source_dir):
>> +            shutil.rmtree(self.source_dir)
> 
>     We do need to track source changes now though ;-)
> 
> I'm not sure what you mean. Are you saying something needs to happen here?

Eventually, when we implement tracking of changes to sources that would
affect build results ;)

Comment (kyrofa) [inline snapcraft/internal/sources.py]: Ah, yes of course. That will be a wonderful day!

File: snapcraft/_baseplugin.py (hunk 2 above)
Comment (kyrofa) [inline snapcraft/_baseplugin.py]: Done.

File: snapcraft/tests/test_sources.py (hunk 3 above)
Comment (kyrofa) [inline snapcraft/tests/test_sources.py]: Done.

**Comment by sergiusens** [comment]
`parts/<part-name>/bin/pkg-config` should be erased from existence :wink:

**Comment by sergiusens** [comment]
just that last comment from me :-)

**Comment by kyrofa** [comment]
> `parts/<part-name>/bin/pkg-config` should be erased from existence

Done :wink: .

**Comment by kyrofa** [comment]
Huh... now that I merged in master suddenly the coverage is decreasing instead of increasing. Not sure what the deal it there. Also, even VPNd in I can't seem to get to the autopkgtest failure, though judging from previous results it's just ROS.

**Comment by kyrofa** [comment]
retest this please

**Comment by kyrofa** [comment]
Huh. Now coverage increased again. Coveralls I think you're full of it.

**Comment by kyrofa** [comment]
retest this please.

**Comment by kyrofa** [comment]
retest this please

**Comment by kyrofa** [comment]
retest this please

**Comment by kyrofa** [comment]
retest this please

**Comment by kyrofa** [comment]
Failing ROS:

```
ValueError: Invalid placeholder in string: line 1, col 123
\x1b[31m\x1b[1m<==\x1b[0m Failed to process package \'\x1b[1m\x1b[34mtalker\x1b[0m\': 
  Invalid placeholder in string: line 1, col 123\x1b[0m
Command failed, exiting.
\x1b[0;31mCommand \'[\'/bin/sh\', \'/tmp/tmpm099_p1z\', \'/bin/bash\', \'/tmp/tmpzbdd1lf3\']\' returned non-zero exit status 1\x1b[0m
'}}}
```

Not really sure what the deal is with autopkgtests...

**Comment by kyrofa** [comment]
retest this please

**Comment by kyrofa** [comment]
retest this please.

**Comment by kyrofa** [comment]
retest this please

**Comment by kyrofa** [comment]
retest this please

**Comment by elopio** [comment]
This is all green. The examples error is in the cleanup, fixed here: https://github.com/ubuntu-core/snappy-jenkins/pull/193

**Pull request merged**